File: lab2.js
const readline = require('readline');

const rl = readline.createInterface({
    input: process.stdin,
    output: process.stdout
});

function minWindow(s, t) {
    if (s.length === 0 || t.length === 0) return "Не найдено";

    const need = {};
    for (let ch of t) {
        need[ch] = (need[ch] || 0) + 1;
    }

    let required = t.length;
    let left = 0;
    let minLen = Infinity;
    let start = 0;

    for (let right = 0; right < s.length; right++) {
        const rc = s[right];
        if (need[rc] !== undefined) {
            if (need[rc] > 0) required--;
            need[rc]--;
        }

        while (required === 0) {
            if (right - left + 1 < minLen) {
                minLen = right - left + 1;
                start = left;
            }

            const lc = s[left];
            if (need[lc] !== undefined) {
                need[lc]++;
                if (need[lc] > 0) required++;
            }
            left++;
        }
    }

    return minLen === Infinity ? "Не найдено" : s.substr(start, minLen);
}

rl.question("Введите строку S: ", (S) => {
    rl.question("Введите строку T: ", (T) => {
        const result = minWindow(S, T);
        console.log(`Результат: ${result}`);
        rl.close();
    });
});

File: lab2.py
def minWindow(s: str, t: str) -> str:
    if not s or not t:
        return "Не найдено"

    need = {}
    for ch in t:
        need[ch] = need.get(ch, 0) + 1

    required = len(t)
    left = 0
    min_len = float('inf')
    start = 0

    for right in range(len(s)):
        rc = s[right]
        if rc in need:
            if need[rc] > 0:
                required -= 1
            need[rc] -= 1

        while required == 0:
            if right - left + 1 < min_len:
                min_len = right - left + 1
                start = left

            lc = s[left]
            if lc in need:
                need[lc] += 1
                if need[lc] > 0:
                    required += 1
            left += 1

    return s[start:start + min_len] if min_len != float('inf') else "Не найдено"

# Ввод от пользователя
S = input("Введите строку S: ")
T = input("Введите строку T: ")

result = minWindow(S, T)
print(f"Результат: {result}")